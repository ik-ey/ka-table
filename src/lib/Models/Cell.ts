export class Cell {
    public columnKey!: string;
    public rowKeyValue!: any;
}
