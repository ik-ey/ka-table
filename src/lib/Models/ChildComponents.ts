import {
    ICellEditorProps,
    ICellProps,
    ICellTextProps,
    IDataRowProps,
    IEmptyCellProps,
    IFilterRowEditorProps,
    IGroupPanelCellProps,
    IGroupPanelProps,
    IGroupRowProps,
    IGroupSummaryCellProps,
    IGroupSummaryRowProps,
    IHeadCellProps,
    IHeadCellResizeProps,
    IHeadRowProps,
    IHeaderFilterButtonProps,
    IHeaderFilterPopupProps,
    ILoadingProps,
    INoDataRowProps,
    IPagingIndexProps,
    IPagingProps,
    IPagingSizeProps,
    IPopupContentItemProps,
    IPopupContentProps,
    IRowProps,
    ISortIconProps,
    ISummaryCellProps,
    ISummaryRowProps,
    ITableBodyProps,
    ITableFootProps,
    ITableHeadProps,
} from '../props';

import { ChildComponent } from './ChildComponent';
import { ITableProps } from '../';

export class ChildComponents<TData= any> {
    public cell?: ChildComponent<ICellProps<TData>>;
    public cellEditor?: ChildComponent<ICellEditorProps<TData>>;
    public cellEditorInput?: ChildComponent<ICellEditorProps<TData>>;
    public cellText?: ChildComponent<ICellTextProps<TData>>;
    public dataAndDetailsRows?: ChildComponent<IRowProps>;
    public dataRow?: ChildComponent<IDataRowProps<TData>>;
    public detailsCell?: ChildComponent<IDataRowProps<TData>>;
    public detailsRow?: ChildComponent<IDataRowProps<TData>>;
    public emptyCell?: ChildComponent<IEmptyCellProps>;
    public filterRowCell?: ChildComponent<IFilterRowEditorProps>;
    public filterRowCellInput?: ChildComponent<IFilterRowEditorProps>;
    public groupCell?: ChildComponent<IGroupRowProps<TData>>;
    public groupExpandButton?: ChildComponent<IGroupRowProps<TData>>;
    public groupPanel?: ChildComponent<IGroupPanelProps>;
    public groupPanelCell?: ChildComponent<IGroupPanelCellProps>;
    public groupRow?: ChildComponent<IGroupRowProps<TData>>;
    public groupSummaryCell?: ChildComponent<IGroupSummaryCellProps<TData>>;
    public groupSummaryRow?: ChildComponent<IGroupSummaryRowProps<TData>>;
    public headCell?: ChildComponent<IHeadCellProps>;
    public headCellContent?: ChildComponent<IHeadCellProps>;
    public headCellResize?: ChildComponent<IHeadCellResizeProps>;
    public headerFilterButton?: ChildComponent<IHeaderFilterButtonProps>;
    public headerFilterPopupContent?: ChildComponent<IPopupContentProps<TData>>;
    public headerFilterPopupRow?: ChildComponent<IDataRowProps>;
    public headerFilterPopupSearchInput?: ChildComponent<IFilterRowEditorProps>;
    public headerFilterPopupTextCell?: ChildComponent<ICellProps>;
    public headRow?: ChildComponent<IHeadRowProps>;
    public loading?: ChildComponent<ILoadingProps>;
    public noDataCell?: ChildComponent<INoDataRowProps>;
    public noDataRow?: ChildComponent<INoDataRowProps>;
    public paging?: ChildComponent<IPagingProps>;
    public pagingIndex?: ChildComponent<IPagingIndexProps>;
    public pagingPages?: ChildComponent<IPagingProps>;
    public pagingSize?: ChildComponent<IPagingSizeProps>;
    public pagingSizes?: ChildComponent<IPagingProps>;
    public rootDiv?: ChildComponent<ITableProps<TData>>;
    public sortIcon?: ChildComponent<ISortIconProps>;
    public summaryCell?: ChildComponent<ISummaryCellProps<TData>>;
    public summaryRow?: ChildComponent<ISummaryRowProps<TData>>;
    public table?: ChildComponent<ITableProps<TData>>;
    public tableBody?: ChildComponent<ITableBodyProps<TData>>;
    public tableFoot?: ChildComponent<ITableFootProps<TData>>;
    public tableHead?: ChildComponent<ITableHeadProps>;
    public tableWrapper?: ChildComponent<ITableProps<TData>>;
}
