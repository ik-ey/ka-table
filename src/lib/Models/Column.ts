import { DataType, SortDirection } from '../enums';

import { Field } from '../types';
import { PopupPosition } from './PopupPosition';

export class Column<TData = any> {
    public colGroup?: React.ColHTMLAttributes<HTMLElement>;
    public dataType?: DataType;
    public field?: Field;
    public filter?: (value: any, filterValue: any, rowData?: any) => boolean;
    public filterRowOperator?: any;
    public filterRowValue?: any;
    public headerFilterListItems?: (props: { data?: TData[] }) => string[];
    public headerFilterPopupPosition?: PopupPosition;
    public headerFilterSearch?: (value: any, searchValue: any, rowData?: any) => boolean;
    public headerFilterSearchValue?: any;
    public headerFilterValues?: string[];
    public isDraggable?: boolean;
    public isEditable?: boolean;
    public isFilterable?: boolean;
    public isHeaderFilterPopupShown?: boolean;
    public isHeaderFilterSearchable?: boolean;
    public isResizable?: boolean;
    public isSortable?: boolean;
    public key!: string;
    public sortDirection?: SortDirection;
    public sortIndex?: number;
    public style?: React.CSSProperties;
    public title?: string;
    public visible?: boolean;
    public width?: number | string;
}
