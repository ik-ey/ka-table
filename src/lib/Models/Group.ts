
export class Group {
    public columnKey!: string;
    public enableSummary?: boolean;
}
