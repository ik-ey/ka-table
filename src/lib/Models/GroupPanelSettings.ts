export class GroupPanelSettings {
    public enabled?: boolean;
    public text?: string;
    public deep?: number;
  }
