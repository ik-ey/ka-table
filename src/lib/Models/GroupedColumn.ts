
export class GroupedColumn {
    public key!: string;
    public title?: string;
    public columnsKeys!: string[];
}
