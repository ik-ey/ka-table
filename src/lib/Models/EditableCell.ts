import { Cell } from './Cell';

export class EditableCell extends Cell {
    public editorValue?: any;
    public validationMessage?: any;
}
