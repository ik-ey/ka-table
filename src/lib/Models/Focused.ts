import { Cell } from './Cell';

export class Focused {
    public cell?: Cell;
    public cellEditorInput?: Cell;
}
