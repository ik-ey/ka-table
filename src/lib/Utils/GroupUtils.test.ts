import {
  convertToFlat,
  getExpandedGroups,
  getGroupMark,
  getGroupText,
  getGroupedStructure,
  groupBy,
  groupSummaryMark,
  isMaxDeep,
  updateExpandedGroups,
} from './GroupUtils';

import { Group } from '../Models/Group';
import { GroupRowData } from '../Models/GroupRowData';

describe('GroupUtils', () => {
  const groupedColumns = [{key: 'country'}, {key: 'type'}];
  const data = [
    { type: 'Cat', name: 'Kas', country: 'Czech Republic' },
    { type: 'Dog', name: 'Rex', country: 'Montenegro' },
    { type: 'Cat', name: 'Simba', country: 'France' },
    { type: 'Dog', name: 'Beethoven', country: 'Czech Republic' },
    { type: 'Cat', name: 'Hash', country: 'Czech Republic' },
  ];
  const groups: Group[] = [{ columnKey: 'country' }, { columnKey: 'type' }];

  describe('convertToFlat', () => {
    it('simple', () => {
      const mappedData = new Map([['France', [{
        country: 'France',
        name: 'Simba',
        type: 'Cat',
      }]]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });

    it('inner group', () => {
      const mappedData = new Map([['France',
        new Map([['Cat', [{
          country: 'France',
          name: 'Simba',
          type: 'Cat',
        }]]]),
      ]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });
    it('inner group with summart', () => {
      const mappedData = new Map([['France',
        new Map([
          ['Cat', [{
            country: 'France',
            name: 'Simba',
            type: 'Cat',
          },
          {
            groupData: [
              {
                country: 'France',
                name: 'Simba',
                type: 'Cat',
              }
            ],
            groupIndex: 0,
            groupSummaryMark,
            key: [['France', 'Cat'], '--: + summary--\\'],
          },
          ]],
          [groupSummaryMark,
            {
              groupData: [
                {
                  country: 'France',
                  name: 'Simba',
                  type: 'Cat',
                }
              ],
              groupIndex: 0,
              groupSummaryMark,
              key: [['France', 'Cat'], '--: + summary--\\'],
            }]
        ]),
      ]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });

    it('complex structure', () => {
      const mappedData = new Map([['France',
        new Map([
          ['Cat', [{
            country: 'France',
            name: 'Simba',
            type: 'Cat',
          }, {
            country: 'France',
            name: 'Kas',
            type: 'Cat',
          }]],
          ['Dog', [{
            country: 'France',
            name: 'Slimu',
            type: 'Dog',
          }, {
            country: 'France',
            name: 'Rex',
            type: 'Dog',
          }]],
        ]),
      ],
      ['Brazil',
        new Map([
          ['Cat', [{
            country: 'Brazil',
            name: 'Simba',
            type: 'Cat',
          }, {
            country: 'Brazil',
            name: 'Kas',
            type: 'Cat',
          }]],
          ['Dog', [{
            country: 'Brazil',
            name: 'Slimu',
            type: 'Dog',
          }, {
            country: 'Brazil',
            name: 'Rex',
            type: 'Dog',
          }]],
        ]),
      ]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });
  });

  it('getExpandedGroups', () => {
    const groupRowData: GroupRowData = { key: ['dog'], value: 'Rex', groupMark: getGroupMark() };
    const groupedData: any[] = [{}, groupRowData, {}];
    const result = getExpandedGroups(groupedData);
    expect(result).toEqual([['dog']]);
  });

  it('updateExpandedGroups add', () => {
    const groupsExpanded: any[][] = [['cat']];
    const updated = updateExpandedGroups(groupsExpanded, ['dog']);
    expect(updated).toEqual([['cat'], ['dog']]);
  });

  it('updateExpandedGroups remove', () => {
    const groupsExpanded: any[][] = [['cat']];
    const updated = updateExpandedGroups(groupsExpanded, ['cat']);
    expect(updated).toEqual([]);
  });

  it('groupBy', () => {
    const result = groupBy(data, (item: any) => item.type);
    expect(result).toMatchSnapshot();
  });

  it('groupBy empty', () => {
    const result = groupBy(data, (item: any) => item.type, true);
    expect(result).toMatchSnapshot();
  });

  describe('getGroupedStructure', () => {
    it('basic', () => {
      const result = getGroupedStructure(data, groups, groupedColumns);
      expect(result).toMatchSnapshot();
    });

    it('expanded root', () => {
      const result = getGroupedStructure(data, groups, groupedColumns, 0, [['Czech Republic']]);
      expect(result).toMatchSnapshot();
    });
    it('expanded second', () => {
      const result = getGroupedStructure(data, groups, groupedColumns, 0, [['Czech Republic'], ['Czech Republic', 'Cat']]);
      expect(result).toMatchSnapshot();
    });
    it('should not expand second', () => {
      const result = getGroupedStructure(data, groups, groupedColumns, 0, [['Czech Republic', 'Cat']]);
      expect(result).toMatchSnapshot();
    });
    it('expanded couple', () => {
      const result = getGroupedStructure(data, groups, groupedColumns, 0,
        [['Czech Republic'], ['Czech Republic', 'Cat'], ['Montenegro']]);
      expect(result).toMatchSnapshot();
    });
    it('expanded couple (skip Czech Republic)', () => {
      const result = getGroupedStructure(data, groups, groupedColumns, 0, [['Czech Republic', 'Cat'], ['Montenegro']]);
      expect(result).toMatchSnapshot();
    });
    it('groupedColumns are empty', () => {
      const result = getGroupedStructure(data, groups, [], 0, [['Czech Republic', 'Cat'], ['Montenegro']]);
      expect(result).toBeUndefined();
    });
    it('returns summary', () => {
      const groupsWithSummary: Group[] = [{ columnKey: 'country', enableSummary: true }, { columnKey: 'type', enableSummary: true }];
      const result = getGroupedStructure(data, groupsWithSummary, groupedColumns, 0);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getGroupText', () => {
    it('default', () => {
      const result = getGroupText('Group Text', { key: 'column1', title: 'Column Title' });
      expect(result).toEqual('Column Title: Group Text');
    });
    it('column.title is not set', () => {
      const result = getGroupText(
        'Group Text',
        { key: 'column1'}
      );
      expect(result).toEqual('Group Text');
    });
    it('format', () => {
      const result = getGroupText(
        'Group Text',
        { key: 'column1', title: 'Column Title' },
        ({column, value}) => `Column: ${column.title}, Value: ${value}`
      );
      expect(result).toEqual('Column: Column Title, Value: Group Text');
    });
  });

  describe('isMaxDeep', () => {
    it('default', () => {
      expect(isMaxDeep({ deep: 1 }, [{key: '1'}, { key: '2'}], [{ columnKey: '1' }])).toBeTruthy();
      expect(isMaxDeep({ deep: 2 }, [{key: '1'}, { key: '2'}], [{ columnKey: '1' }])).toBeFalsy();
      expect(isMaxDeep({ }, [{key: '1'}, { key: '2'}, { key: '3'}], [{ columnKey: '1' }])).toBeFalsy();
      expect(isMaxDeep({ }, [{key: '1'}, { key: '2'}], [{ columnKey: '1' }])).toBeTruthy();
      expect(isMaxDeep({ }, [{key: '1'}], [{ columnKey: '1' }])).toBeTruthy();
      expect(isMaxDeep({ deep: 1 }, [{key: '1'}, { key: '2'}])).toBeFalsy();
    });
  });
});
