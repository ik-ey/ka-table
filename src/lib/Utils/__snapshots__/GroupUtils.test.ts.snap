// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupUtils convertToFlat complex structure 1`] = `
Array [
  Object {
    "groupItems": Map {
      "Cat" => Array [
        Object {
          "country": "France",
          "name": "Simba",
          "type": "Cat",
        },
        Object {
          "country": "France",
          "name": "Kas",
          "type": "Cat",
        },
      ],
      "Dog" => Array [
        Object {
          "country": "France",
          "name": "Slimu",
          "type": "Dog",
        },
        Object {
          "country": "France",
          "name": "Rex",
          "type": "Dog",
        },
      ],
    },
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "groupItems": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
      Object {
        "country": "France",
        "name": "Kas",
        "type": "Cat",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "France",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
  Object {
    "country": "France",
    "name": "Kas",
    "type": "Cat",
  },
  Object {
    "groupItems": Array [
      Object {
        "country": "France",
        "name": "Slimu",
        "type": "Dog",
      },
      Object {
        "country": "France",
        "name": "Rex",
        "type": "Dog",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "France",
      "Dog",
    ],
    "value": "Dog",
  },
  Object {
    "country": "France",
    "name": "Slimu",
    "type": "Dog",
  },
  Object {
    "country": "France",
    "name": "Rex",
    "type": "Dog",
  },
  Object {
    "groupItems": Map {
      "Cat" => Array [
        Object {
          "country": "Brazil",
          "name": "Simba",
          "type": "Cat",
        },
        Object {
          "country": "Brazil",
          "name": "Kas",
          "type": "Cat",
        },
      ],
      "Dog" => Array [
        Object {
          "country": "Brazil",
          "name": "Slimu",
          "type": "Dog",
        },
        Object {
          "country": "Brazil",
          "name": "Rex",
          "type": "Dog",
        },
      ],
    },
    "groupMark": Object {},
    "key": Array [
      "Brazil",
    ],
    "value": "Brazil",
  },
  Object {
    "groupItems": Array [
      Object {
        "country": "Brazil",
        "name": "Simba",
        "type": "Cat",
      },
      Object {
        "country": "Brazil",
        "name": "Kas",
        "type": "Cat",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "Brazil",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "Brazil",
    "name": "Simba",
    "type": "Cat",
  },
  Object {
    "country": "Brazil",
    "name": "Kas",
    "type": "Cat",
  },
  Object {
    "groupItems": Array [
      Object {
        "country": "Brazil",
        "name": "Slimu",
        "type": "Dog",
      },
      Object {
        "country": "Brazil",
        "name": "Rex",
        "type": "Dog",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "Brazil",
      "Dog",
    ],
    "value": "Dog",
  },
  Object {
    "country": "Brazil",
    "name": "Slimu",
    "type": "Dog",
  },
  Object {
    "country": "Brazil",
    "name": "Rex",
    "type": "Dog",
  },
]
`;

exports[`GroupUtils convertToFlat inner group 1`] = `
Array [
  Object {
    "groupItems": Map {
      "Cat" => Array [
        Object {
          "country": "France",
          "name": "Simba",
          "type": "Cat",
        },
      ],
    },
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "groupItems": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "France",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
]
`;

exports[`GroupUtils convertToFlat inner group with summart 1`] = `
Array [
  Object {
    "groupItems": Map {
      "Cat" => Array [
        Object {
          "country": "France",
          "name": "Simba",
          "type": "Cat",
        },
        Object {
          "groupData": Array [
            Object {
              "country": "France",
              "name": "Simba",
              "type": "Cat",
            },
          ],
          "groupIndex": 0,
          "groupSummaryMark": Object {},
          "key": Array [
            Array [
              "France",
              "Cat",
            ],
            "--: + summary--\\\\",
          ],
        },
      ],
      Object {} => Object {
        "groupData": Array [
          Object {
            "country": "France",
            "name": "Simba",
            "type": "Cat",
          },
        ],
        "groupIndex": 0,
        "groupSummaryMark": Object {},
        "key": Array [
          Array [
            "France",
            "Cat",
          ],
          "--: + summary--\\\\",
        ],
      },
    },
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "groupItems": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
      Object {
        "groupData": Array [
          Object {
            "country": "France",
            "name": "Simba",
            "type": "Cat",
          },
        ],
        "groupIndex": 0,
        "groupSummaryMark": Object {},
        "key": Array [
          Array [
            "France",
            "Cat",
          ],
          "--: + summary--\\\\",
        ],
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "France",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
  Object {
    "groupData": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
    "groupIndex": 0,
    "groupSummaryMark": Object {},
    "key": Array [
      Array [
        "France",
        "Cat",
      ],
      "--: + summary--\\\\",
    ],
  },
  Object {
    "groupData": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
    "groupIndex": 0,
    "groupSummaryMark": Object {},
    "key": Array [
      Array [
        "France",
        "Cat",
      ],
      "--: + summary--\\\\",
    ],
  },
]
`;

exports[`GroupUtils convertToFlat simple - group items 1`] = `
Array [
  Object {
    "groupItems": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
]
`;

exports[`GroupUtils convertToFlat simple 1`] = `
Array [
  Object {
    "groupItems": Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
]
`;

exports[`GroupUtils getGroupedStructure basic 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
    ],
    "Dog" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Beethoven",
        "type": "Dog",
      },
    ],
  },
  "Montenegro" => Map {
    "Dog" => Array [
      Object {
        "country": "Montenegro",
        "name": "Rex",
        "type": "Dog",
      },
    ],
  },
  "France" => Map {
    "Cat" => Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
  },
}
`;

exports[`GroupUtils getGroupedStructure expanded couple (skip Czech Republic) 1`] = `
Map {
  "Czech Republic" => Array [],
  "Montenegro" => Map {
    "Dog" => Array [],
  },
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure expanded couple 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
    ],
    "Dog" => Array [],
  },
  "Montenegro" => Map {
    "Dog" => Array [],
  },
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure expanded root 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [],
    "Dog" => Array [],
  },
  "Montenegro" => Array [],
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure expanded second 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
    ],
    "Dog" => Array [],
  },
  "Montenegro" => Array [],
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure returns summary 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
      Object {
        "groupData": Array [
          Object {
            "country": "Czech Republic",
            "name": "Kas",
            "type": "Cat",
          },
          Object {
            "country": "Czech Republic",
            "name": "Hash",
            "type": "Cat",
          },
        ],
        "groupIndex": 1,
        "groupSummaryMark": Object {},
        "key": "[[\\"Czech Republic\\",\\"Cat\\"],\\"--:+summary--\\"]",
      },
    ],
    "Dog" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Beethoven",
        "type": "Dog",
      },
      Object {
        "groupData": Array [
          Object {
            "country": "Czech Republic",
            "name": "Beethoven",
            "type": "Dog",
          },
        ],
        "groupIndex": 1,
        "groupSummaryMark": Object {},
        "key": "[[\\"Czech Republic\\",\\"Dog\\"],\\"--:+summary--\\"]",
      },
    ],
    Object {} => Object {
      "groupData": Array [
        Object {
          "country": "Czech Republic",
          "name": "Kas",
          "type": "Cat",
        },
        Object {
          "country": "Czech Republic",
          "name": "Beethoven",
          "type": "Dog",
        },
        Object {
          "country": "Czech Republic",
          "name": "Hash",
          "type": "Cat",
        },
      ],
      "groupIndex": 0,
      "groupSummaryMark": Object {},
      "key": "[[\\"Czech Republic\\"],\\"--:+summary--\\"]",
    },
  },
  "Montenegro" => Map {
    "Dog" => Array [
      Object {
        "country": "Montenegro",
        "name": "Rex",
        "type": "Dog",
      },
      Object {
        "groupData": Array [
          Object {
            "country": "Montenegro",
            "name": "Rex",
            "type": "Dog",
          },
        ],
        "groupIndex": 1,
        "groupSummaryMark": Object {},
        "key": "[[\\"Montenegro\\",\\"Dog\\"],\\"--:+summary--\\"]",
      },
    ],
    Object {} => Object {
      "groupData": Array [
        Object {
          "country": "Montenegro",
          "name": "Rex",
          "type": "Dog",
        },
      ],
      "groupIndex": 0,
      "groupSummaryMark": Object {},
      "key": "[[\\"Montenegro\\"],\\"--:+summary--\\"]",
    },
  },
  "France" => Map {
    "Cat" => Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
      Object {
        "groupData": Array [
          Object {
            "country": "France",
            "name": "Simba",
            "type": "Cat",
          },
        ],
        "groupIndex": 1,
        "groupSummaryMark": Object {},
        "key": "[[\\"France\\",\\"Cat\\"],\\"--:+summary--\\"]",
      },
    ],
    Object {} => Object {
      "groupData": Array [
        Object {
          "country": "France",
          "name": "Simba",
          "type": "Cat",
        },
      ],
      "groupIndex": 0,
      "groupSummaryMark": Object {},
      "key": "[[\\"France\\"],\\"--:+summary--\\"]",
    },
  },
}
`;

exports[`GroupUtils getGroupedStructure should not expand second 1`] = `
Map {
  "Czech Republic" => Array [],
  "Montenegro" => Array [],
  "France" => Array [],
}
`;

exports[`GroupUtils groupBy 1`] = `
Map {
  "Cat" => Array [
    Object {
      "country": "Czech Republic",
      "name": "Kas",
      "type": "Cat",
    },
    Object {
      "country": "France",
      "name": "Simba",
      "type": "Cat",
    },
    Object {
      "country": "Czech Republic",
      "name": "Hash",
      "type": "Cat",
    },
  ],
  "Dog" => Array [
    Object {
      "country": "Montenegro",
      "name": "Rex",
      "type": "Dog",
    },
    Object {
      "country": "Czech Republic",
      "name": "Beethoven",
      "type": "Dog",
    },
  ],
}
`;

exports[`GroupUtils groupBy empty 1`] = `
Map {
  "Cat" => Array [],
  "Dog" => Array [],
}
`;
